### DISCUSSION TITLE
Does RestoreUseStaticGraphEvaluation require /graph

### DISCUSSION THREAD

**Discussion by stan-sz** [dotnet/msbuild#8326] [Q&A] [answered]
I have a case in a larger repo where `RestoreUseStaticGraphEvaluation=false` and `/graph` is enabled.
In one of the projects `dotnet restore` does not honor the `RuntimeIdentifier` and does not create the right entry in `project.assets.json` for `net6.0/win-x64`. A subsequent `dotnet build --no-restore` fails on missing assets for that `TF/RI` for the project.

Setting `RestoreUseStaticGraphEvaluation=true` and `/graph` solves the problem.

Should `RestoreUseStaticGraphEvaluation` and `/graph` features go hand-in-hand @rainersigwald and @jeffkl?

**Comment by jeffkl** [comment, marked as answer]
Short answer: No

Long answer:

Traditionally, MSBuild command-line based builds build from the **top** down.  What this means is that the first project being built is started and for each reference, a new build is spawned.  Each spawned build waits for the dependencies until they are built.  For large builds, this means you could have hundreds of projects waiting.  This is done by starting the build of the entry project, gathering the `<ProjectReference />` items, pausing that thread, and scheduling a new build of each project reference via the `<MSBuild />` task.

Static graph in MSBuild instead evaluates all of the projects ahead of time, then schedules the builds from the **bottom** up.  This means that no projects are scheduled and paused which is more efficient.  The builds are scheduled in dependency order starting at the bottom, 
 and maximizing the parallelism based on dependency order.  However, there are a few breaking changes in the way builds are scheduled if you have heavily customized your build so it can't be the default.  You opt into using static graph based evaluation and **build scheduling** with the `/graph` command-line argument.

Static graph-based NuGet restore only uses the MSBuild API that loads all of the projects ahead of time but doesn't use it for build scheduling.  The static graph API is just very efficient for loading projects and then the information is read to run a restore in NuGet.  But this does not use any of the build scheduling functionality that really makes up `/graph`.

**Comment by stan-sz** [comment]
Just in case linking to https://github.com/NuGet/Home/issues/9803
